File: python/calendar.py
import datetime
import locale
import sys

# ロケールを日本語に設定
locale.setlocale(locale.LC_TIME, 'ja_JP.UTF-8')


# -mで指定した月をreturnする
# -mで引数を指定しない場合は今月をreturnする
def get_month(argv):
    idx = find_dash_m(argv)
    # -mが存在しない、または-mのインデックスがリストの最後の要素の場合、今月をreturnする
    if idx is None or idx == len(argv)-1:
        return datetime.date.today().month
    # -mで指定された値が有効な値の場合、その月をreturnする
    if is_valid_dash_m(argv, idx):
        return int(argv[idx+1])


# -mで指定した値が文字列もしくは不正な月の場合はエラーメッセージを出力してプログラムを終了させる
# -mで指定した値が有効の場合はTrueを返す
def is_valid_dash_m(argv, idx):
    try:
        dash_m_value = int(argv[idx+1])
        if 1 <= dash_m_value <= 12:
            return True
        else:
            raise ValueError
    except ValueError:
        print(f'{argv[idx+1]} is neither a month number (1..12) nor a name')
        sys.exit(1)


# コマンドライン引数のリストの中から-mがあるインデックスを戻す
def find_dash_m(argv):
    if '-m' in argv:
        return argv.index('-m')
    else:
        return None


# コマンドライン引数のリストを受け取る
argv = sys.argv

# 西暦、月、曜日を定義
today = datetime.date.today()
year = today.year
month = get_month(argv)
weekdays = ['月', '火', '水', '木', '金', '土', '日']

# 指定月のdateインスタンスを生成
date = datetime.date(today.year, month, 1)

# 出力用にフォーマットする
formatted_date = date.strftime('%B %Y')
space_fill = 20
formatted_weekdays = ' '.join(weekdays)

# 月と西暦を中央寄せで表示
print(f'{formatted_date:^{space_fill}}')

# 曜日を月曜日から順に表示
print(formatted_weekdays)

# 月の最終日は何日か計算
if month == 12:
    next_month = datetime.date(year+1, 1, 1)
else:
    next_month = datetime.date(year, month%12+1, 1)
last_day = (next_month - date).days

# 指定した月の1日は何曜日か計算
weekday_of_first_day = date.weekday()

# 1日から31日までを表示
for i, v in enumerate(range(1,last_day+1)):
    # 1日目の曜日に合わせて空白を入れる
    if i == 0:
        print('   ' * weekday_of_first_day, end='')

    # 今日の日付をハイライトする
    if v == today.day:
        print('\033[48;2;204;204;204m'+f'{v:2d}'+'\033[0m', end='')
    else:
        print(f'{v:2d}', end='')

    # iが日曜日、vが最終日の場合、改行を入れる
    # そうでない場合、空白を入れる
    if i % 7 == (6 - weekday_of_first_day) or v == last_day:
        print('')
    else:
        print(' ', end='')

File: javascript/calendar.js
// 曜日を定義
const weekDay = ["日", "月", "火", "水", "木", "金", "土"];

// 現在時刻を取得
const now = new Date();

// 年月日、その日の曜日をそれぞれ取得
const nowYear = now.getFullYear();
let month = now.getMonth();

// -mのコマンドオプションの処理
mOfIndex = process.argv.indexOf("-m");

// -mがコマンドオプションに含まれている場合
if (mOfIndex !== -1) {
  const choicedMonth = process.argv.at(mOfIndex + 1);

  // -mに続く引数がない時
  if (choicedMonth === undefined) {
    // 何もしない処理
  } else if (Number(choicedMonth) >= 1 && Number(choicedMonth) <= 12) {
    // 引数が存在する場合
    // 引数が1以上12以下の時、monthを更新
    month = Number(choicedMonth) - 1;
  } else {
    // 引数が不正の値のとき、エラーメッセージを出力
    console.log("月の値が不正です。\n正しい月を入力してください。");
    process.exit(1);
  }
}

// カレンダーの西暦と月を出力
const calendarHeader = `${month + 1}月 ${nowYear}`;
const spacing = "      ";
console.log(spacing + calendarHeader + spacing);

// 曜日を出力
console.log(weekDay.join(" "));

// 日付を出力

// 月末の日付を取得
const totalDays = new Date(nowYear, month + 1, 0).getDate();

// 1日から月末までの数字が入った配列を定義
const days = [];
for (let i = 1; i <= totalDays; i++) {
  days.push(String(i).padStart(2, " "));
}

// 曜日のインデックスと1日の曜日を合わせるため空白を追加
const firstDayOfTheWeek = new Date(nowYear, month, 1).getDay();
for (let i = 0; i < firstDayOfTheWeek; i++) {
  days.unshift("  ");
}

// 日付を出力
for (let i = 0; i < days.length; i += 7) {
  const weekDays = days.slice(i, i + 7);
  console.log(weekDays.join(" "));
}
